### PR TITLE
LCE tests and `SolidMechanics` initial guess helper

### PR DESCRIPTION
**Pull request by barreracruz**
Wrote two small but representative tests for the two liquid crystal elastomer models being used for an LDRD-ER.

### CODE REVIEW

**Comment by jamiebramwell** [comment]
Now that the nonlinear solver PR is merged, this should be good to go.

File: src/serac/physics/base_physics.hpp [hunk 1]
```diff
@@ -310,6 +310,9 @@ class BasePhysics {
     /// The finite element states representing user-defined and owned parameter fields
     serac::FiniteElementState* state;
 
+    /// The finite element state representing the parameter at the previous evalutaion
```
Comment (white238) [inline src/serac/physics/base_physics.hpp]:
```suggestion
    /// The finite element state representing the parameter at the previous evaluation
```

File: src/serac/physics/materials/liquid_crystal_elastomer.hpp [hunk 2]
```diff
@@ -46,8 +48,8 @@ struct LiquidCrystalElastomer {
    * @param transition_temperature Characteristic temperature of the order-disorder transition
    * @param N_b_squared Number of Kunh segments/chain, times square of Kuhn segment length
    */
-  LiquidCrystalElastomer(double rho, double shear_modulus, double bulk_modulus, double order_constant,
-                         double order_parameter, double transition_temperature, double N_b_squared)
+  LiqCrystElast_Brighenti(double rho, double shear_modulus, double bulk_modulus, double order_constant,
```
Comment (white238) [inline src/serac/physics/materials/liquid_crystal_elastomer.hpp]: My vote is for `LiquidCrystalElastomer_Brighenti` but not a hill i'll die on.

File: src/serac/physics/materials/liquid_crystal_elastomer.hpp [hunk 3]
```diff
@@ -69,20 +71,25 @@ struct LiquidCrystalElastomer {
...
    * @param[in,out] state A state variable object for this material. The value is updated in place.
    * @param[in] displacement_grad displacement gradient with respect to the reference configuration
    * @param[in] temperature the temperature
-   * @param[in] gamma the polar angle used to define the liquid crystal orientation vector
+   * @param[in] gamma the first polar angle used to define the liquid crystal orientation vector
+   * @param[in] eta the secpmd polar angle used to define the liquid crystal orientation vector
    * @return The calculated material response (Cauchy stress) for the material
    */
-  template <typename DispGradType, typename TemperatureType, typename AngleType>
+  template <typename DispGradType, typename orderParamType, typename GammaAngleType>
```
Comment (white238) [inline src/serac/physics/materials/liquid_crystal_elastomer.hpp]: The second template parameter's doxygen needs to be updated.

File: src/serac/physics/base_physics.hpp [hunk 4]
```diff
@@ -310,6 +310,9 @@ class BasePhysics {
     /// The finite element states representing user-defined and owned parameter fields
     serac::FiniteElementState* state;
 
+    /// The finite element state representing the parameter at the previous evaluation
+    std::unique_ptr<serac::FiniteElementState> old_state;
```
Comment (white238) [inline src/serac/physics/base_physics.hpp]: `previous_state`?

File: src/serac/physics/materials/liquid_crystal_elastomer.hpp [hunk 5]
```diff
@@ -179,4 +184,144 @@ struct LiquidCrystalElastomer {
...
+   * @tparam GammaAngleType number-like type for the orientation angle gamma
+   *
+   * @param[in,out] state A state variable object for this material. The value is updated in place.
+   * @param[in] displacement_grad displacement gradient with respect to the reference configuration
+   * @param[in] inst_order_param_tuple the current order parameter
+   * @param[in] gamma the first polar angle used to define the liquid crystal orientation vector
+   * @param[in] eta the secpmd polar angle used to define the liquid crystal orientation vector
+   * @return The calculated material response (Cauchy stress) for the material
+   */
+  template <typename DispGradType, typename orderParamType, typename GammaAngleType, typename EtaAngleType>
```
Comment (white238) [inline src/serac/physics/materials/liquid_crystal_elastomer.hpp]: tparam's are wrong

File: src/serac/physics/materials/liquid_crystal_elastomer.hpp [hunk 6]
```diff
@@ -179,4 +184,144 @@ struct LiquidCrystalElastomer {
...
+   *
+   * @tparam DisplacementType number-like type for the displacement vector
+   * @tparam DispGradType number-like type for the displacement gradient tensor
+   * @tparam GammaAngleType number-like type for the orientation angle gamma
+   *
+   * @param[in,out] state A state variable object for this material. The value is updated in place.
+   * @param[in] displacement_grad displacement gradient with respect to the reference configuration
+   * @param[in] inst_order_param_tuple the current order parameter
+   * @param[in] gamma the first polar angle used to define the liquid crystal orientation vector
+   * @param[in] eta the secpmd polar angle used to define the liquid crystal orientation vector
```
Comment (white238) [inline src/serac/physics/materials/liquid_crystal_elastomer.hpp]:
```suggestion
   * @param[in] gamma_tuple the first polar angle used to define the liquid crystal orientation vector
   * @param[in] eta_tuple the second polar angle used to define the liquid crystal orientation vector
```

File: src/serac/physics/materials/liquid_crystal_elastomer.hpp [hunk 7]
```diff
@@ -179,4 +184,144 @@ struct LiquidCrystalElastomer {
...
+  /// -------------------------------------------------------
+
+  /**
+   * @brief Material response
+   *
+   * @tparam DisplacementType number-like type for the displacement vector
+   * @tparam DispGradType number-like type for the displacement gradient tensor
+   * @tparam GammaAngleType number-like type for the orientation angle gamma
+   *
+   * @param[in,out] state A state variable object for this material. The value is updated in place.
```
Comment (white238) [inline src/serac/physics/materials/liquid_crystal_elastomer.hpp]: Apparently this is an open feature request on Doxygen, but should we leave unnamed variables here or remove them?

File: src/serac/physics/materials/liquid_crystal_elastomer.hpp [hunk 8]
```diff
@@ -179,4 +184,144 @@ struct LiquidCrystalElastomer {
...
+
+    // transform from second Piola-Lichhoff to Cauchy stress
+    auto stress = 1.0 / J * F * (S_stress_1 + S_stress_2 + S_stress_3) * transpose(F);
+
+    return stress;
+  }
+
+  /// -------------------------------------------------------
+
+  template <typename DispGradType, typename orderParamType, typename GammaAngleType, typename EtaAngleType>
```
Comment (white238) [inline src/serac/physics/materials/liquid_crystal_elastomer.hpp:311]: Missing documentation.

File: src/serac/physics/state/state_manager.cpp [hunk 9]
```diff
@@ -257,7 +257,6 @@ mfem::ParMesh* StateManager::setMesh(std::unique_ptr<mfem::ParMesh> pmesh, const
   // 3. Uses the spatial dimension as the mesh dimension (i.e. it is not a lower dimension manifold)
   // 4. Uses nodal instead of VDIM ordering (i.e. xxxyyyzzz instead of xyzxyzxyz)
   pmesh->SetCurvature(1, is_discontinuous, -1, mfem::Ordering::byNODES);
-
```
Comment (white238) [inline src/serac/physics/state/state_manager.cpp]: This should be reverted.

File: src/serac/physics/tests/lce_Bertoldi_lattice.cpp [hunk 10]
```diff
@@ -0,0 +1,184 @@
+// Copyright (c) 2019-2022, Lawrence Livermore National Security, LLC and
```
Comment (white238) [inline src/serac/physics/tests/lce_Bertoldi_lattice.cpp]:
```suggestion
// Copyright (c) 2019-2023, Lawrence Livermore National Security, LLC and
```

File: src/serac/physics/tests/lce_Bertoldi_lattice.cpp [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+  }
+
+  // check output
+#ifdef PERIODIC_MESH
+  EXPECT_NEAR(gblDispYmin, -2.27938e-05, 1.0e-6);
+#else
+  EXPECT_NEAR(gblDispYmin, -2.92599e-05, 1.0e-6);
+#endif
+  serac::exitGracefully();
+}
```
Comment (white238) [inline src/serac/physics/tests/lce_Bertoldi_lattice.cpp]:
```suggestion
}

```

File: src/serac/physics/tests/lce_Brighenti_tensile.cpp [hunk 12]
```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2019-2022, Lawrence Livermore National Security, LLC and
```
Comment (white238) [inline src/serac/physics/tests/lce_Brighenti_tensile.cpp]:
```suggestion
// Copyright (c) 2019-2023, Lawrence Livermore National Security, LLC and
```

File: src/serac/physics/tests/lce_Brighenti_tensile.cpp [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+    // update pseudotime-dependent information
+    t += dt;
+    loadVal = iniLoadVal + (maxLoadVal - iniLoadVal) * std::pow(t / tmax, 0.75);
+  }
+
+  // check output
+  EXPECT_NEAR(gblDispYmax, 1.95036097e-05, 1.0e-8);
+
+  serac::exitGracefully();
+}
```
Comment (white238) [inline src/serac/physics/tests/lce_Brighenti_tensile.cpp]:
```suggestion
}

```

File: src/serac/physics/tests/lce_Brighenti_tensile.cpp [hunk 14]
```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2019-2022, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include <fstream>
+
+#include "axom/slic/core/SimpleLogger.hpp"
```
Comment (white238) [inline src/serac/physics/tests/lce_Brighenti_tensile.cpp]:
```suggestion
```

File: src/serac/physics/tests/lce_Brighenti_tensile.cpp [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+#include "serac/serac_config.hpp"
+#include "serac/mesh/mesh_utils.hpp"
+#include "serac/physics/state/state_manager.hpp"
+#include "serac/physics/solid_mechanics.hpp"
+#include "serac/physics/materials/liquid_crystal_elastomer.hpp"
+#include "serac/infrastructure/initialize.hpp"
+#include "serac/infrastructure/terminator.hpp"
+
+using namespace serac;
+
```
Comment (white238) [inline src/serac/physics/tests/lce_Brighenti_tensile.cpp:21]:
```suggestion
```

File: src/serac/physics/tests/lce_Brighenti_tensile.cpp [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+              std::pow(x[0] - 1.0, 2) + std::pow(x[1] - 1.0, 2) - std::pow(rad, 2) < 0.0)
+                 ? 0.333 * M_PI_2
+                 : 0.667 * M_PI_2;
+    }
+  };
+
+  mfem::FunctionCoefficient coef(gamma_func);
+  gamma.project(coef);
+
+  // Construct a functional-based solid mechanics solver
```
Comment (white238) [inline src/serac/physics/tests/lce_Brighenti_tensile.cpp]:
```suggestion
  // Construct a solid mechanics solver
```

File: src/serac/physics/tests/lce_Brighenti_tensile.cpp [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+      pmesh);
+
+  Functional<double(H1<p, dim>)> area({&solid_solver.displacement().space()});
+  area.AddSurfaceIntegral(
+      DependsOn<>{}, [=](auto x, auto /*n*/) { return (x[1] > 0.99 * ly) ? 1.0 : 0.0; }, pmesh);
+
+  double initial_area = area(solid_solver.displacement());
+  if (rank == 0) {
+    std::cout << "... Initial Area of the top surface: " << initial_area << std::endl;
+  }
```
Comment (white238) [inline src/serac/physics/tests/lce_Brighenti_tensile.cpp]:
```suggestion
  SLIC_INFO_ROOT("... Initial Area of the top surface: " << initial_area);
```

File: src/serac/physics/tests/lce_Brighenti_tensile.cpp [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+
+  // Perform remaining quasi-static solve
+  for (int i = 0; i < (num_steps + 1); i++) {
+    if (rank == 0) {
+      std::cout << "\n\n............................"
+                << "\n... Entering time step: " << i + 1 << "\n............................\n"
+                << "\n... At time: " << t << "\n... And with a tension load of: " << loadVal << " ("
+                << loadVal / maxLoadVal * 100 << "`%` of max)"
+                << "\n... And with uniform temperature of: " << initial_temperature << std::endl;
+    }
```
Comment (white238) [inline src/serac/physics/tests/lce_Brighenti_tensile.cpp]:
```suggestion
    SLIC_INFO_ROOT("\n\n............................"
                << "\n... Entering time step: " << i + 1 << "\n............................\n"
                << "\n... At time: " << t << "\n... And with a tension load of: " << loadVal << " ("
                << loadVal / maxLoadVal * 100 << "`%` of max)"
                << "\n... And with uniform temperature of: " << initial_temperature);
```

File: src/serac/physics/tests/lce_Brighenti_tensile.cpp [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+                  << current_qoi / current_area << std::endl;
+      }
+
+      if (std::isnan(gblDispYmax)) {
+        if (rank == 0) {
+          std::cout << "... Solution blew up... Check boundary and initial conditions." << std::endl;
+        }
+        exit(1);
+      }
+    }
```
Comment (white238) [inline src/serac/physics/tests/lce_Brighenti_tensile.cpp]:
```suggestion
      SLIC_ERROR_ROOT_IF(std::isnan(gblDispYmax), "... Solution blew up... Check boundary and initial conditions.");
```

File: src/serac/physics/materials/liquid_crystal_elastomer.hpp (hunk 7 above)
Comment (jamiebramwell) [reply src/serac/physics/materials/liquid_crystal_elastomer.hpp]: I say remove them.

**Comment by jamiebramwell** [comment]
I think this is now ready.

File: src/serac/physics/materials/liquid_crystal_elastomer.hpp [hunk 20]
```diff
@@ -179,4 +197,159 @@ struct LiquidCrystalElastomer {
...
+ * @brief Bertoldi's liquid crystal elastomer model
+ * Paper: Li, S., Librandi, G., Yao, Y., Richard, A. J., Schneider‐Yamamura, A., Aizenberg, J., & Bertoldi, K. (2021).
+ * Controlling Liquid Crystal Orientations for Programmable Anisotropic Transformations in Cellular Microstructures.
+ * Advanced Materials, 33(42), 2105024.
+ */
+struct LiquidCrystalElastomerBertoldi {
+  /// this model is only intended to be used in 3D
+  static constexpr int dim = 3;
+
+  /// internal variables for the liquid crystal elastomer model
```
Comment (white238) [inline src/serac/physics/materials/liquid_crystal_elastomer.hpp]: Can you add a note as to why this is empty?

**Review by white238 — APPROVED** [review]

**Review by samuelpmishLLNL — APPROVED** [review]

**Pull request merged**